File: app/src/main/java/forpdateam/ru/forpda/views/messagepanel/MessagePanelBehavior.java
package forpdateam.ru.forpda.views.messagepanel;

import android.support.design.widget.AppBarLayout;
import android.support.design.widget.CoordinatorLayout;
import android.support.v7.widget.CardView;
import android.view.View;

import forpdateam.ru.forpda.App;

/**
 * Created by radiationx on 07.01.17.
 */

public class MessagePanelBehavior extends CoordinatorLayout.Behavior<CardView> {
    private boolean canScrolling = true;

    public MessagePanelBehavior() {
        super();
    }

    @Override
    public boolean onStartNestedScroll(final CoordinatorLayout coordinatorLayout, final CardView child,
                                       final View directTargetChild, final View target, final int nestedScrollAxes) {
        if (!canScrolling)
            child.setTranslationY(0);
        return canScrolling;
    }


    public void setCanScrolling(boolean canScrolling) {
        this.canScrolling = canScrolling;
    }


    @Override
    public boolean layoutDependsOn(CoordinatorLayout parent, CardView child, View dependency) {
        return dependency instanceof AppBarLayout;
    }

    @Override
    public boolean onDependentViewChanged(CoordinatorLayout parent, CardView child, View dependency) {
        if (!canScrolling) return false;
        float percent = 1.0f - ((float) -dependency.getTop() / (float) dependency.getMeasuredHeight());
        int scrolled = (int) ((child.getMeasuredHeight() + (2 * App.px8)) * percent);
        child.setTranslationY(scrolled);
        return true;
    }
}

File: app/src/main/java/forpdateam/ru/forpda/ui/views/messagepanel/MessagePanelBehavior.kt
package forpdateam.ru.forpda.ui.views.messagepanel

import android.content.Context
import android.view.View
import androidx.cardview.widget.CardView
import androidx.coordinatorlayout.widget.CoordinatorLayout
import com.google.android.material.appbar.AppBarLayout
import forpdateam.ru.forpda.R
import forpdateam.ru.forpda.extensions.getDimenPx

/**
 * Created by radiationx on 07.01.17.
 */
class MessagePanelBehavior(private val context: Context) : CoordinatorLayout.Behavior<CardView>() {
    private var canScrolling = true
    private val dp8 = context.getDimenPx(R.dimen.dp8)
    override fun onStartNestedScroll(
        coordinatorLayout: CoordinatorLayout, child: CardView,
        directTargetChild: View, target: View, nestedScrollAxes: Int
    ): Boolean {
        if (!canScrolling) child.translationY = 0f
        return canScrolling
    }


    fun setCanScrolling(canScrolling: Boolean) {
        this.canScrolling = canScrolling
    }


    override fun layoutDependsOn(
        parent: CoordinatorLayout,
        child: CardView,
        dependency: View
    ): Boolean {
        return dependency is AppBarLayout
    }

    override fun onDependentViewChanged(
        parent: CoordinatorLayout,
        child: CardView,
        dependency: View
    ): Boolean {
        if (!canScrolling) return false
        val percent = 1.0f - (-dependency.top.toFloat() / dependency.measuredHeight.toFloat())

        val scrolled = ((child.measuredHeight + (2 * dp8)) * percent).toInt()
        child.translationY = scrolled.toFloat()
        return true
    }
}
